File: 54-spiral-matrix/54-spiral-matrix.cpp
class Solution{
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix){
        vector<int> ans;

        int top=0, bottom=matrix.size()-1, left=0, right=matrix[0].size()-1;

        int dir=1;

        while(top <= bottom and left <= right){

            if(dir == 1){
                // left to right print
                for(int i=left; i<=right; i++){
                    ans.push_back(matrix[top][i]);
                }
                top++;
                dir = 2;
            }

            else if(dir == 2){

                // top to bottom
                for(int i=top; i<=bottom; i++){
                    ans.push_back(matrix[i][right]);
                }
                right--;
                dir = 3;
            }

            else if(dir == 3){

                // right to left
                for(int i=right; i>=left; i--){
                    ans.push_back(matrix[bottom][i]);
                }
                bottom--;
                dir = 4;
            }
            
            else if(dir == 4){
                // bottom to up
                for(int i=bottom; i>= top; i--){
                    ans.push_back(matrix[i][left]);
                }
                left++;
                dir = 1;
            }
        }

        return ans;
    }
};

File: 54-spiral-matrix/54-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        
        output = []
        top, bottom, left, right = 0, len(matrix)-1, 0, len(matrix[0])-1
        
        direction = 1
        
        while top <= bottom and left <= right:
            
            if direction == 1:
                # left to right
                for i in range(left, right+1):
                    output.append(matrix[top][i])
                
                top += 1
                direction = 2
            
            elif direction == 2:
                # top to bottom
                for i in range(top, bottom+1):
                    output.append(matrix[i][right])
                
                right -= 1
                direction = 3
            
            elif direction == 3:
                # right to left
                for i in range(right, left-1, -1):
                    output.append(matrix[bottom][i])
                
                bottom -= 1
                direction = 4
            
            else:
                # bottom to top
                for i in range(bottom, top-1, -1):
                    output.append(matrix[i][left])
                
                left += 1
                direction = 1
                
        return output
    